File: EIDSS v5/android.java/workspace/EIDSS/src/com/bv/eidss/data/EidssDatabase.java
package com.bv.eidss.data;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import com.WSParser.WebServices.EidssService.VectorBaseReferenceRaw;
import com.WSParser.WebServices.EidssService.VectorBaseReferenceTranslationRaw;
import com.WSParser.WebServices.EidssService.VectorGisBaseReferenceRaw;
import com.WSParser.WebServices.EidssService.VectorGisBaseReferenceTranslationRaw;
import com.bv.eidss.model.BaseReference;
import com.bv.eidss.model.GisBaseReference;
import com.bv.eidss.model.HumanCase;
import com.bv.eidss.model.Options;


public class EidssDatabase extends SQLiteOpenHelper {

    private static final String create_sql_options = 
"CREATE TABLE Options\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", strServerUrl TEXT NULL\n" +
", strLocalPassword TEXT NULL\n" +
", strLoginOrganization TEXT NULL\n" +
", strLoginUsername TEXT NULL\n" +
")";
    private static final String create_sql_human_case =
"CREATE TABLE HumanCase\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", strLastSynError TEXT NULL\n" +
", intStatus INT NULL\n" +
", datCreateDate DATE NULL\n" +
"\n" +
", uidOfflineCaseID TEXT NULL\n" +
", strCaseID TEXT NULL\n" +
", idfCase BIGINT NULL\n" +
"\n" +
", strLocalIdentifier TEXT NULL\n" +
", idfsTentativeDiagnosis BIGINT NULL\n" +
", datTentativeDiagnosisDate DATE NULL\n" +
", strFamilyName TEXT NULL\n" +
", strFirstName TEXT NULL\n" +
", datDateofBirth DATE NULL\n" +
", intPatientAge INT NULL\n" +
", idfsHumanAgeType BIGINT NULL\n" +
", idfsHumanGender BIGINT NULL\n" +
", idfsRegionCurrentResidence BIGINT NULL\n" +
", idfsRayonCurrentResidence BIGINT NULL\n" +
", idfsSettlementCurrentResidence BIGINT NULL\n" +
", strBuilding TEXT NULL\n" +
", strHouse TEXT NULL\n" +
", strApartment TEXT NULL\n" +
", strStreetName TEXT NULL\n" +
", strPostCode TEXT NULL\n" +
", strHomePhone TEXT NULL\n" +
", datOnSetDate DATE NULL\n" +
", idfsFinalState BIGINT NULL\n" +
", idfsHospitalizationStatus BIGINT NULL\n" +
", datNotificationDate DATE NULL\n" +
", strSentByOffice TEXT NULL\n" +
", strSentByPerson TEXT NULL\n" +
")";
    private static final String create_sql_references = 
"CREATE TABLE BaseReference\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", idfsBaseReference BIGINT NULL\n" +
", idfsReferenceType BIGINT NULL\n" +
", intHACode INT NULL\n" +
", strDefault TEXT NULL\n" +
")";
    private static final String create_sql_gis_references = 
"CREATE TABLE GisBaseReference\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", idfsBaseReference BIGINT NULL\n" +
", idfsReferenceType BIGINT NULL\n" +
", idfsCountry BIGINT NULL\n" +
", idfsRegion BIGINT NULL\n" +
", idfsRayon BIGINT NULL\n" +
", strDefault TEXT NULL\n" +
")";
    private static final String create_sql_references_lang = 
"CREATE TABLE BaseReferenceTranslation\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", idfsBaseReference BIGINT NULL\n" +
", strTranslation TEXT NULL\n" +
", strLanguage TEXT NULL\n" +
")";
    private static final String create_sql_gis_references_lang = 
"CREATE TABLE GisBaseReferenceTranslation\n" +
"(\n" +
"  id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE\n" +
", idfsBaseReference BIGINT NULL\n" +
", strTranslation TEXT NULL\n" +
", strLanguage TEXT NULL\n" +
")";
    private static final String insert_options_def =
/*    		
"INSERT INTO Options\n" +
"(\n" +
"strServerUrl,\n" +
"strLocalPassword,\n" +
"strLoginOrganization,\n" +
"strLoginUsername\n" +
")\n" +
"VALUES\n" +
"(\n" +
"'http://95.167.107.83:8083/',\n" +
"'',\n" +
"'molhsa',\n" +
"'demo'\n" +
")";
*/
"INSERT INTO Options\n" +
"(\n" +
"strServerUrl,\n" +
"strLocalPassword,\n" +
"strLoginOrganization,\n" +
"strLoginUsername\n" +
")\n" +
"VALUES\n" +
"(\n" +
"'http://oa.eidss.info/',\n" +
//"'http://oa.eidss.or.tz/',\n" +
//"'https://192.168.10.34:1443/',\n" +
//"'http://192.168.10.34:8080/',\n" +
"'',\n" +
//"'NCDC&PH',\n" +
"'',\n" +
//"'test_admin'\n" +
"''\n" +
")";

    private static final String select_sql_reference = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM BaseReference br\n" +
"LEFT OUTER JOIN BaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = ?\n" +
"ORDER BY name";
    private static final String select_sql_reference_hacodable = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM BaseReference br\n" +
"LEFT OUTER JOIN BaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = ? and br.intHACode & ?\n" +
"ORDER BY name";
    private static final String select_sql_country = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM GisBaseReference br\n" +
"LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = 19000001\n" +
"ORDER BY name";
    private static final String select_sql_regions = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM GisBaseReference br\n" +
"LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = 19000003\n" +
"ORDER BY name";
    private static final String select_sql_rayons = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM GisBaseReference br\n" +
"LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = 19000002 and (br.idfsRegion = ? or br.idfsBaseReference = 0)\n" +
"ORDER BY name";
    private static final String select_sql_settlements = 
"SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name\n" +
"FROM GisBaseReference br\n" +
"LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?\n" +
"WHERE br.idfsReferenceType = 19000004 and (br.idfsRayon = ? or br.idfsBaseReference = 0)\n" +
"ORDER BY name";
	
    private final static String DatabaseName = "eidss.db";
    private final static int DatabaseVersion = 1;  
    public EidssDatabase(Context context) {
    	super(context, DatabaseName, null, DatabaseVersion);
    }
	
	@Override
	public void onCreate(SQLiteDatabase db) {
        db.execSQL(create_sql_options);
        db.execSQL(insert_options_def);
        db.execSQL(create_sql_references);
        db.execSQL(create_sql_references_lang);
        db.execSQL(create_sql_gis_references);
        db.execSQL(create_sql_gis_references_lang);
        db.execSQL(create_sql_human_case);
	}

	@Override
	public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
	}
	
    public String getCurrentLanguage()
    {
    	String lang = Locale.getDefault().getLanguage();
    	String country = Locale.getDefault().getCountry();
        if (lang.compareTo("ru") == 0)
            return "ru";
        if (lang.compareTo("en") == 0){
            if (country.compareTo("AU") == 0){
                return "ka";
            }
            if (country.compareTo("GB") == 0){
                return "ka";
            }
        }
    	return "en";
    }
	

    public Options Options()
    {
    	Cursor cursor = getReadableDatabase().query("Options", null, null, null, null, null, null);
        cursor.moveToFirst();
        Options ret = Options.FromCursor(cursor);
        return ret;
    }
    public Options OptionsUpdate(Options o)
    {
        getWritableDatabase().replace("Options", null, o.ContentValues());
        return o;
    }
    
    public boolean LoadReferences(VectorBaseReferenceRaw refs, VectorBaseReferenceTranslationRaw refs_trans, 
    		VectorGisBaseReferenceRaw gis_refs, VectorGisBaseReferenceTranslationRaw gis_refs_trans)
    {
    	try{
	    	SQLiteDatabase db = getWritableDatabase();
	    	db.beginTransaction();
	    	db.delete("GisBaseReferenceTranslation", null, null);
	    	db.delete("GisBaseReference", null, null);
	    	db.delete("BaseReferenceTranslation", null, null);
	    	db.delete("BaseReference", null, null);
	    	
	    	for(int i = 0; i < refs.size(); i++){
	    		db.insert("BaseReference", null, refs.get(i).ContentValues());
	    	}
	    	for(int i = 0; i < refs_trans.size(); i++){
	    		db.insert("BaseReferenceTranslation", null, refs_trans.get(i).ContentValues());
	    	}
	    	for(int i = 0; i < gis_refs.size(); i++){
	    		db.insert("GisBaseReference", null, gis_refs.get(i).ContentValues());
	    	}
	    	for(int i = 0; i < gis_refs_trans.size(); i++){
	    		db.insert("GisBaseReferenceTranslation", null, gis_refs_trans.get(i).ContentValues());
	    	}
	    	
	        db.setTransactionSuccessful();
	        db.endTransaction();
	        return true;
    	}
    	catch(Exception e){
            e.printStackTrace();
    		return false;
    	}
    }

    public List<BaseReference> Reference(long type, String lang, int hacode)
    {
    	List<BaseReference> ret = new ArrayList<BaseReference>();
    	String sql;
    	String[] args;
    	if (hacode == 0) {
    		sql = select_sql_reference;
    		args = new String[2];
    		args[0] = lang;
    		args[1] = String.valueOf(type);
    	}
    	else {
    		sql = select_sql_reference_hacodable;
    		args = new String[3];
    		args[0] = lang;
    		args[1] = String.valueOf(type);
    		args[2] = String.valueOf(hacode);
    	}
    	Cursor cursor = getReadableDatabase().rawQuery(sql, args);
        Boolean r = cursor.moveToFirst();
        while (r)
        {
            ret.add(BaseReference.FromCursor(cursor));
            r = cursor.moveToNext();
        }
        cursor.close();
        //ret.Sort((a, b) => a.name.CompareTo(b.name));
        return ret;
    }
    
    public GisBaseReference GisCountry(String lang)
    {
        GisBaseReference ret = null;
    	String[] args = new String[1];
    	args[0] = lang;
    	Cursor cursor = getReadableDatabase().rawQuery(select_sql_country, args);
        Boolean r = cursor.moveToFirst();
        while (r)
        {
        	ret = GisBaseReference.FromCursor(cursor);
            r = cursor.moveToNext();
        }
        cursor.close();
        return ret;
    }
    public List<GisBaseReference> GisRegions(String lang)
    {
    	List<GisBaseReference> ret = new ArrayList<GisBaseReference>();
    	String[] args = new String[1];
    	args[0] = lang;
    	Cursor cursor = getReadableDatabase().rawQuery(select_sql_regions, args);
        Boolean r = cursor.moveToFirst();
        while (r)
        {
            ret.add(GisBaseReference.FromCursor(cursor));
            r = cursor.moveToNext();
        }
        cursor.close();
        //ret.Sort((a, b) => a.name.CompareTo(b.name));
        return ret;
    }
    public List<GisBaseReference> GisRayons(long region, String lang)
    {
    	List<GisBaseReference> ret = new ArrayList<GisBaseReference>();
    	String[] args = new String[2];
    	args[0] = lang;
    	args[1] = String.valueOf(region);
    	Cursor cursor = getReadableDatabase().rawQuery(select_sql_rayons, args);
        Boolean r = cursor.moveToFirst();
        while (r)
        {
            ret.add(GisBaseReference.FromCursor(cursor));
            r = cursor.moveToNext();
        }
        cursor.close();
        //ret.Sort((a, b) => a.name.CompareTo(b.name));
        return ret;
    }
    public List<GisBaseReference> GisSettlements(long rayon, String lang)
    {
    	List<GisBaseReference> ret = new ArrayList<GisBaseReference>();
    	String[] args = new String[2];
    	args[0] = lang;
    	args[1] = String.valueOf(rayon);
    	Cursor cursor = getReadableDatabase().rawQuery(select_sql_settlements, args);
        Boolean r = cursor.moveToFirst();
        while (r)
        {
            ret.add(GisBaseReference.FromCursor(cursor));
            r = cursor.moveToNext();
        }
        cursor.close();
        //ret.Sort((a, b) => a.name.CompareTo(b.name));
        return ret;
    }

    public List<HumanCase> HumanCaseSelect(int status)
    {
    	List<HumanCase> ret = new ArrayList<HumanCase>();
    	Cursor cursor = getReadableDatabase().query("HumanCase", null, null, null, null, null, "datCreateDate");
    	Boolean r = cursor.moveToFirst();
    	while(r)
    	{
    		HumanCase hc = HumanCase.FromCursor(cursor); 
    		if (hc != null && (status == 0 || hc.getStatus() == status)) 
    			ret.add(hc);
    		r = cursor.moveToNext();
    	}
    	cursor.close();
        return ret;
    }
        
    public HumanCase HumanCaseInsert(HumanCase hc)
    {
        hc.setId(getWritableDatabase().insert("HumanCase", null, hc.ContentValues()));
        return hc;
    }
    public HumanCase HumanCaseUpdate(HumanCase hc)
    {
    	getWritableDatabase().replace("HumanCase", null, hc.ContentValues());
        return hc;
    }
    public void HumanCaseDelete(HumanCase hc)
    {
    	String[] ids = new String[1]; ids[0] = String.valueOf(hc.getId());
    	getWritableDatabase().delete("HumanCase", "id = ?", ids);
    }

}

File: EIDSS v5/android/eidssdroid/Database/EidssDatabase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using Android.App;
using Android.Content;
using Android.Database;
using Android.Database.Sqlite;
using Android.OS;
using Android.Runtime;
using Android.Views;
using Android.Widget;
using bv.com.eidss;
using eidssdroid.Model;

namespace eidssdroid.model.Database
{
    public class EidssDatabase : SQLiteOpenHelper
    {
        private static readonly string create_sql_options = @"
CREATE TABLE Options
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , strServerUrl TEXT NULL
    , strLocalPassword TEXT NULL
    , strLoginOrganization TEXT NULL
    , strLoginUsername TEXT NULL
)";
        private static readonly string create_sql_human_case = @"
CREATE TABLE HumanCase
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , strLastSynError TEXT NULL
    , intStatus INT NULL
    , datCreateDate DATE NULL

    , uidOfflineCaseID TEXT NULL
    , strCaseID TEXT NULL
    , idfCase BIGINT NULL

    , strLocalIdentifier TEXT NULL
    , idfsTentativeDiagnosis BIGINT NULL
    , datTentativeDiagnosisDate DATE NULL
    , strFamilyName TEXT NULL
    , strFirstName TEXT NULL
    , datDateofBirth DATE NULL
    , intPatientAge INT NULL
    , idfsHumanAgeType BIGINT NULL
    , idfsHumanGender BIGINT NULL
    , idfsRegionCurrentResidence BIGINT NULL
    , idfsRayonCurrentResidence BIGINT NULL
    , idfsSettlementCurrentResidence BIGINT NULL
    , strBuilding TEXT NULL
    , strHouse TEXT NULL
    , strApartment TEXT NULL
    , strStreetName TEXT NULL
    , strPostCode TEXT NULL
    , strHomePhone TEXT NULL
    , datOnSetDate DATE NULL
    , idfsFinalState BIGINT NULL
    , idfsHospitalizationStatus BIGINT NULL
    , datNotificationDate DATE NULL
    , strSentByOffice TEXT NULL
    , strSentByPerson TEXT NULL
)";
        private static readonly string create_sql_references = @"
CREATE TABLE BaseReference
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , idfsBaseReference BIGINT NULL
    , idfsReferenceType BIGINT NULL
    , intHACode INT NULL
    , strDefault TEXT NULL
)";
        private static readonly string create_sql_gis_references = @"
CREATE TABLE GisBaseReference
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , idfsBaseReference BIGINT NULL
    , idfsReferenceType BIGINT NULL
    , idfsCountry BIGINT NULL
    , idfsRegion BIGINT NULL
    , idfsRayon BIGINT NULL
    , strDefault TEXT NULL
)";
        private static readonly string create_sql_references_lang = @"
CREATE TABLE BaseReferenceTranslation
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , idfsBaseReference BIGINT NULL
    , strTranslation TEXT NULL
    , strLanguage TEXT NULL
)";
        private static readonly string create_sql_gis_references_lang = @"
CREATE TABLE GisBaseReferenceTranslation
(
      id INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE
    , idfsBaseReference BIGINT NULL
    , strTranslation TEXT NULL
    , strLanguage TEXT NULL
)";
        private static readonly string insert_options_def = @"
INSERT INTO Options
(
    strServerUrl,
    strLocalPassword,
    strLoginOrganization,
    strLoginUsername
)
VALUES
(
    'http://192.168.10.34:8080/',
    '',
    'NCDC&PH',
    'test_admin'
)
";
        private static readonly string select_sql_reference = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM BaseReference br
LEFT OUTER JOIN BaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = ?
";
        private static readonly string select_sql_reference_hacodable = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM BaseReference br
LEFT OUTER JOIN BaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = ? and br.intHACode & ?
";
        private static readonly string select_sql_country = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM GisBaseReference br
LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = 19000001
";
        private static readonly string select_sql_regions = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM GisBaseReference br
LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = 19000003
";
        private static readonly string select_sql_rayons = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM GisBaseReference br
LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = 19000002 and (br.idfsRegion = ? or br.idfsBaseReference = 0)
";
        private static readonly string select_sql_settlements = @"
SELECT br.idfsBaseReference as idfsBaseReference, ifnull(bt.strTranslation, br.strDefault) as name
FROM GisBaseReference br
LEFT OUTER JOIN GisBaseReferenceTranslation bt on bt.idfsBaseReference = br.idfsBaseReference and bt.strLanguage = ?
WHERE br.idfsReferenceType = 19000004 and (br.idfsRayon = ? or br.idfsBaseReference = 0)
";

        
        private static new readonly string DatabaseName = "eidss.db";
        private static readonly int DatabaseVersion = 1;  
        public EidssDatabase(Context context) : base(context, DatabaseName, null, DatabaseVersion) {}

        public override void OnCreate(SQLiteDatabase db)
        {
            db.ExecSQL(create_sql_options);
            db.ExecSQL(insert_options_def);
            db.ExecSQL(create_sql_references);
            db.ExecSQL(create_sql_references_lang);
            db.ExecSQL(create_sql_gis_references);
            db.ExecSQL(create_sql_gis_references_lang);
            db.ExecSQL(create_sql_human_case);
        }

        public override void OnUpgrade(SQLiteDatabase db, int oldVersion, int newVersion)
        {   // not required until second version :)
            throw new NotImplementedException();
        }

        public string CurrentLanguage
        {
            get
            {
                if (Thread.CurrentThread.CurrentUICulture.TwoLetterISOLanguageName.ToLower() == "ru")
                    return "ru";
                if (Thread.CurrentThread.CurrentUICulture.Name.ToLower() == "en-au")
                    return "ka";
                if (Thread.CurrentThread.CurrentUICulture.Name.ToLower() == "en-gb")
                    return "ka";
                return "en";
            }
        }


        public Options Options()
        {
            using(var cursor = ReadableDatabase.Query("Options", null, null, null, null, null, null))
            {
                cursor.MoveToFirst();
                var ret = eidssdroid.Model.Options.FromCursor(cursor);
                return ret;
            }
        }
        public Options OptionsUpdate(Options o)
        {
            WritableDatabase.Replace("Options", null, o.ContentValues());
            return o;
        }

        public bool LoadReferences(BaseReferenceRaw[] refs, BaseReferenceTranslationRaw[] refs_trans, 
            GisBaseReferenceRaw[] gis_refs, GisBaseReferenceTranslationRaw[] gis_refs_trans)
        {
            try
            {
                WritableDatabase.BeginTransaction();
                WritableDatabase.Delete("GisBaseReferenceTranslation", null, null);
                WritableDatabase.Delete("GisBaseReference", null, null);
                WritableDatabase.Delete("BaseReferenceTranslation", null, null);
                WritableDatabase.Delete("BaseReference", null, null);

                foreach (BaseReferenceRaw c in refs.GroupBy(c => c.idfsReferenceType).Select(
                            g => new BaseReferenceRaw()
                            {idfsBaseReference = 0, idfsReferenceType = g.Key, intHACode = -1, strDefault = ""}))
                    WritableDatabase.Insert("BaseReference", null, c.ContentValues());
                foreach (BaseReferenceRaw c in refs)
                    WritableDatabase.Insert("BaseReference", null, c.ContentValues());

                /*foreach (BaseReferenceTranslationRaw c in refs_trans.GroupBy(c => c.strLanguage).Select(
                            g => new BaseReferenceTranslationRaw() { idfsBaseReference = 0, strLanguage = g.Key, strTranslation = ""}))
                    WritableDatabase.Insert("BaseReferenceTranslation", null, c.ContentValues());*/
                foreach (BaseReferenceTranslationRaw c in refs_trans)
                    WritableDatabase.Insert("BaseReferenceTranslation", null, c.ContentValues());

                foreach (GisBaseReferenceRaw c in gis_refs.GroupBy(c => c.idfsReferenceType).Select(
                            g => new GisBaseReferenceRaw() { idfsBaseReference = 0, idfsReferenceType = g.Key, strDefault = "" }))
                    WritableDatabase.Insert("GisBaseReference", null, c.ContentValues());
                foreach (GisBaseReferenceRaw c in gis_refs)
                    WritableDatabase.Insert("GisBaseReference", null, c.ContentValues());

                /*foreach (GisBaseReferenceTranslationRaw c in gis_refs_trans.GroupBy(c => c.strLanguage).Select(
                            g => new GisBaseReferenceTranslationRaw() { idfsBaseReference = 0, strLanguage = g.Key, strTranslation = "" }))
                    WritableDatabase.Insert("BaseReferenceTranslation", null, c.ContentValues());*/
                foreach (GisBaseReferenceTranslationRaw c in gis_refs_trans)
                    WritableDatabase.Insert("GisBaseReferenceTranslation", null, c.ContentValues());

                WritableDatabase.SetTransactionSuccessful();
                WritableDatabase.EndTransaction();
                return true;
            }
            catch (Exception/* ex */)
            {
                WritableDatabase.EndTransaction();
                return false;
            }
        }

        public List<BaseReference> Reference(long type, string lang, int hacode)
        {
            var ret = new List<BaseReference>();
            using (var cursor = ReadableDatabase.RawQuery(
                hacode == 0 ? select_sql_reference : select_sql_reference_hacodable,
                hacode == 0 ? new[] { lang, type.ToString() } : new[] { lang, type.ToString(), hacode.ToString() }))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret.Add(BaseReference.FromCursor(cursor));
                    r = cursor.MoveToNext();
                }
                ret.Sort((a, b) => a.name.CompareTo(b.name));
                return ret;
            }
        }
        public GisBaseReference GisCountry(string lang)
        {
            GisBaseReference ret = null;
            using (var cursor = ReadableDatabase.RawQuery(select_sql_country, new[] { lang }))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret = GisBaseReference.FromCursor(cursor);
                    r = cursor.MoveToNext();
                }
            }
            return ret;
        }
        public List<GisBaseReference> GisRegions(string lang)
        {
            var ret = new List<GisBaseReference>();
            using (var cursor = ReadableDatabase.RawQuery(select_sql_regions, new[] { lang }))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret.Add(GisBaseReference.FromCursor(cursor));
                    r = cursor.MoveToNext();
                }
                ret.Sort((a, b) => a.name.CompareTo(b.name));
                return ret;
            }
        }
        public List<GisBaseReference> GisRayons(long region, string lang)
        {
            var ret = new List<GisBaseReference>();
            using (var cursor = ReadableDatabase.RawQuery(select_sql_rayons, new[] { lang, region.ToString() }))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret.Add(GisBaseReference.FromCursor(cursor));
                    r = cursor.MoveToNext();
                }
                ret.Sort((a, b) => a.name.CompareTo(b.name));
                return ret;
            }
        }
        public List<GisBaseReference> GisSettlements(long rayon, string lang)
        {
            var ret = new List<GisBaseReference>();
            using (var cursor = ReadableDatabase.RawQuery(select_sql_settlements, new[] { lang, rayon.ToString() }))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret.Add(GisBaseReference.FromCursor(cursor));
                    r = cursor.MoveToNext();
                }
                ret.Sort((a, b) => a.name.CompareTo(b.name));
                return ret;
            }
        }


        public List<HumanCase> HumanCaseSelect()
        {
            var ret = new List<HumanCase>();
            using (var cursor = ReadableDatabase.Query("HumanCase", null, null, null, null, null, null))
            {
                var r = cursor.MoveToFirst();
                while (r)
                {
                    ret.Add(HumanCase.FromCursor(cursor));
                    r = cursor.MoveToNext();
                }
                ret.Sort((a, b) => b.datCreateDate.CompareTo(a.datCreateDate));
                return ret;
            }
        }
        public HumanCase HumanCaseInsert(HumanCase hc)
        {
            hc.id = WritableDatabase.Insert("HumanCase", null, hc.ContentValues());
            return hc;
        }
        public HumanCase HumanCaseUpdate(HumanCase hc)
        {
            WritableDatabase.Replace("HumanCase", null, hc.ContentValues());
            return hc;
        }
        public void HumanCaseDelete(HumanCase hc)
        {
            WritableDatabase.Delete("HumanCase", "id = ?", new [] {hc.id.ToString()});
        }
    }

}

